### PR TITLE
Handle hoisted Catch local in state machine rewriter

### LINKED ISSUE #863: VB assert failure: "captured local should not need rewriting"
Assert failure compiling the following with vbc.exe /t:library c.vb

Imports System
Imports System.Threading.Tasks
Class C
    Shared Function F() As Object
        Throw New ArgumentException()
    End Function
    Shared Async Function M() As Task
        Dim o As Object
        Try
            o = F()
        Catch e As Exception
            o = e
        End Try
    End Function
End Class

### PR DESCRIPTION
**Pull request by cston**
Fixes #863

### CODE REVIEW

**Comment by cston** [comment]
@gafter, @AlekseyTs

**Comment by gafter** [comment]
Is there some bug that this is fixing?

**Comment by cston** [comment]
Fixes #863

**Comment by gafter** [comment]
:+1:

File: src/Compilers/VisualBasic/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.StateMachineMethodToClassRewriter.vb [hunk 1]
```diff
@@ -329,28 +329,39 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
-                    Debug.Assert(Not Me.Proxies.ContainsKey(origLocal), "captured local should not need rewriting")
+            Private Function TryRewriteLocal(local As LocalSymbol) As LocalSymbol
+                If NeedsProxy(local) Then
+                    ' no longer a local symbol
+                    Return Nothing
+                End If
 
-                    Dim newType = VisitType(origLocal.Type)
-                    If newType = origLocal.Type Then
+                Dim newLocal As LocalSymbol = Nothing
```
Comment (VladimirReshetnikov) [inline src/Compilers/VisualBasic/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.StateMachineMethodToClassRewriter.vb]: The initializer 'Nothing' is not used on any control path and looks unnecessary.

**Pull request merged**